src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1994,6 +2024,129 @@ IntersectPlacementList(List *lhsPlacementList, List *rhsPlacementList)
 
 
 /*
+ * GroupInsertValuesByShardId takes care of grouping the rows from a multi-row
+ * INSERT by target shard. At this point, all pruning has taken place and we
+ * need only to build sets of rows for each destination. This is done by a
+ * simple sort (by shard identifier) and gather step. The sort has the side-
+ * effect of getting things in ascending order to avoid unnecessary deadlocks
+ * during Task execution.
+ */
+static List *
+GroupInsertValuesByShardId(List *insertValuesList)
+{
+	ModifyRoute *route = NULL;
+	ListCell *insertValuesCell = NULL;
+	List *modifyRouteList = NIL;
+
+	insertValuesList = SortList(insertValuesList, CompareInsertValuesByShardId);
```

Comment:
Is that a good idea? `qsort`, which is what `SortList` uses, isn't guaranteed to be stable - in fact . Therefore sorting like this has a good chance of removing the ordering of individual rows, which in turn can cause problems because the ordering might have been intention for deadlock avoidance purposes.
Seems better to have a `shardId->List hashmap`, a `shardOffset` indexed array of lists, or a `bsearch()`able array (not list!) of shards (think I'd go with the last variant), that things get split over. As that makes the loop body simpler, I don't think you'll even end up with more code.  As an aside, it avoids the ridiculous practice of turning a list into an array, then back into a list ;)